packages/core/src/stateUtils.ts

```diff
@@ -505,55 +496,20 @@ function resolveHistory<TContext, TEvent extends EventObject>(
         )
       : getInitialStateNodes(parent);
   }
-  const subHistoryValue = nestedPath<HistoryValue>(parent.path, 'states')(
-    historyValue
-  ).current;
-  if (isString(subHistoryValue)) {
-    return [getStateNode(parent, subHistoryValue)];
-  }
-  return flatten(
-    toStatePaths(subHistoryValue!).map(subStatePath => {
-      return stateNode.history === 'deep'
-        ? getFromRelativePath(parent, subStatePath)
-        : [parent.states[subStatePath[0]]];
-    })
-  );
+
+  return state.historyMap[stateNode.id];
```

Comment:
I don't immediately see why this matters here as `state.historyMap` seems to always be checked beforehand, but running tests without has confirmed that this is very much needed. Do you maybe know, from the top of your head, why exactly this is needed?

Reply:
This is needed for determining initial state nodes (see `getFromRelativePath()` calls).